pkg/controller/portworxdiag/portworxdiag.go

```diff
@@ -107,8 +138,394 @@ func (c *Controller) Reconcile(ctx context.Context, req reconcile.Request) (reco
+		if diag.Status.Message != "" {
+			op = "replace"
+		}
+		patches = append(patches, map[string]interface{}{
+			"op":    op,
+			"path":  "/status/message",
+			"value": newMessage,
+		})
+	}
+
+	return patches
+}
+
+func getMissingNodeStatusesPatch(diag *diagv1.PortworxDiag, nodeStatusesToAdd []*diagv1.NodeStatus) []map[string]interface{} {
+	if len(nodeStatusesToAdd) == 0 {
+		return nil
+	}
+	patches := []map[string]interface{}{}
+	if diag.Status.NodeStatuses == nil {
+		patches = append(patches, map[string]interface{}{
+			"op":    "add",
+			"path":  "/status/nodes",
+			"value": []diagv1.NodeStatus{},
+		})
+	}
+	for _, toAdd := range nodeStatusesToAdd {
+		patches = append(patches, map[string]interface{}{
+			"op":    "add",
+			"path":  "/status/nodes/-",
+			"value": toAdd,
+		})
+	}
+	return patches
+}
+
+func (c *Controller) updateDiagFields(diag *diagv1.PortworxDiag, stc *corev1.StorageCluster, prs *podReconcileStatus) error {
+	patches := []map[string]interface{}{}
+
+	if patch := getMissingStatusPatch(diag); patch != nil {
+		patches = append(patches, patch)
+	}
+
+	if patch := getChangedClusterUUIDPatch(diag, stc); patch != nil {
+		patches = append(patches, patch)
+	}
+
+	if phasePatches := getOverallPhasePatch(diag); len(phasePatches) > 0 {
+		patches = append(patches, phasePatches...)
+	}
+
+	if phasePatches := getMissingNodeStatusesPatch(diag, prs.nodeStatusesToAdd); len(phasePatches) > 0 {
+		patches = append(patches, phasePatches...)
+	}
+
+	body, err := json.Marshal(patches)
+	if err != nil {
+		return fmt.Errorf("failed to marshal json patch to JSON: %v", err)
+	}
+
+	err = c.client.Status().Patch(context.TODO(), diag, client.RawPatch(types.JSONPatchType, body))
+	if err != nil {
+		return fmt.Errorf("failed to update phase for PortworxDiag CR: %v", err)
+	}
+	return nil
+}
+
+func (c *Controller) patchPhase(diag *diagv1.PortworxDiag, newPhase string, newMessage string) error {
+	patch := []map[string]interface{}{
+		{
+			"op":    "add",
+			"path":  "/status/phase",
+			"value": newPhase,
+		},
+		{
+			"op":    "add",
+			"path":  "/status/message",
+			"value": newMessage,
+		},
+	}
+	body, err := json.Marshal(patch)
+	if err != nil {
+		return fmt.Errorf("failed to marshal json patch to JSON: %v", err)
+	}
+	err = c.client.Status().Patch(context.TODO(), diag, client.RawPatch(types.JSONPatchType, body))
+	if err != nil {
+		return fmt.Errorf("failed to update phase for PortworxDiag CR: %v", err)
+	}
+	return nil
+}
+
+func (c *Controller) syncPortworxDiag(diag *diagv1.PortworxDiag) error {
+	// If diag is already done, don't do any more work
+	switch diag.Status.Phase {
+	case diagv1.DiagStatusPartialFailure:
+		fallthrough
+	case diagv1.DiagStatusCompleted:
+		fallthrough
+	case diagv1.DiagStatusFailed:
+		logrus.Infof("PortworxDiag %s in namespace %s is already in status %s, no work to do", diag.Name, diag.Namespace, diag.Status.Phase)
+		return nil
+	default:
+		break
+	}
+
+	logrus.Info("Enter syncPortworxDiag")
```

Comment:
nit: do we still want this as Info? Wondering if this could be spammy. However, if `syncPortworxDiag` is only called on a diag CR change/etc, this should be fine.

Reply:
It's called only on changes, or on a 30 second interval I think. However, we have a fast exit for already completed/failed ones, so it's only spammy while the Diag object is still running :wink:

I'll leave it as is while we're still early in development but if it ends up being spammy we can change it later (I have lots more work to do in these files)